tool/tsh/aws.go

```diff
@@ -0,0 +1,310 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509/pkix"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/gravitational/trace"
+	"github.com/pborman/uuid"
+
+	"github.com/gravitational/teleport/lib/asciitable"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	alpncommon "github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+
+	awsarn "github.com/aws/aws-sdk-go/aws/arn"
+)
+
+const (
+	awsCLIBinaryName = "aws"
+)
+
+func onAWS(cf *CLIConf) error {
+	// create self-signed local cert AWS LocalProxy listener cert
+	// and pass CA to AWS CLI by --ca-bundle flag to enforce HTTPS
+	// protocol communication between AWS CLI <-> LocalProxy internal.
+	tmpCert, err := newTempSelfSignedLocalCert()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if err := tmpCert.Clean(); err != nil {
+			log.WithError(err).Errorf(
+				"Failed to clean temporary self-signed local proxy cert %q.", tmpCert.getCAPath())
+		}
+	}()
+
+	// ENV AWS credentials need to be set in order to enforce AWS CLI to
+	// sign the request and provide Authorization Header where service-name and region-name are encoded.
+	// When endpoint-url AWS CLI flag provides the destination AWS API address is override by endpoint-url value.
+	// Teleport AWS Signing APP will resolve aws-service and aws-region to the proper Amazon API URL.
+	generatedAWSCred, err := genAndSetAWSCredentials()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tc, err := makeClient(cf, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	lp, err := createLocalAWSCLIProxy(cf, tc, generatedAWSCred, tmpCert.getCert())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer lp.Close()
+	go func() {
+		if err := lp.StartAWSAccessProxy(cf.Context); err != nil {
+			log.WithError(err).Errorf("Failed to start local proxy.")
+		}
+	}()
+
+	addr, err := utils.ParseAddr(lp.GetAddr())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	url := url.URL{
+		Path:   "/",
+		Host:   fmt.Sprintf("%s:%d", "localhost", addr.Port(0)),
+		Scheme: "https",
+	}
+
+	endpointFlag := fmt.Sprintf("--endpoint-url=%s", url.String())
+	bundleFlag := fmt.Sprintf("--ca-bundle=%s", tmpCert.getCAPath())
+
+	args := append([]string{}, cf.AWSCommandArgs...)
+	args = append(args, endpointFlag)
+	args = append(args, bundleFlag)
+	cmd := exec.Command(awsCLIBinaryName, args...)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	if err := cmd.Run(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func genAndSetAWSCredentials() (*credentials.Credentials, error) {
```

Comment:
Nit: Let's please add a comment clarifying this function's purpose (wrt fake creds generation).